pkg/awsutils/awsutils.go

```diff
@@ -404,13 +404,49 @@ func (cache *EC2InstanceMetadataCache) refreshSGIDs(mac string) error {
 	addedSGs := newSGs.Difference(&cache.securityGroups)
 	deletedSGs := cache.securityGroups.Difference(&newSGs)
 
+	addedSGsCount := 0
+	deletedSGsCount := 0
+
 	for _, sg := range addedSGs.SortedList() {
 		log.Infof("Found %s, added to ipamd cache", sg)
+		addedSGsCount++
 	}
 	for _, sg := range deletedSGs.SortedList() {
 		log.Infof("Removed %s from ipamd cache", sg)
+		deletedSGsCount++
 	}
+
+	log.Infof("AddedSGsCount = %d and DeletedSgsCount = %d", addedSGsCount, deletedSGsCount)
+
 	cache.securityGroups.Set(sgIDs)
+
+	if addedSGsCount != 0 || deletedSGsCount != 0 {
+		var sgIDsPtrs []*string
+		sgIDsPtrs = aws.StringSlice(sgIDs)
+
+		// Fetch all local ENI info from metadata
+		allENIs, err := cache.GetAttachedENIs()
+		if err != nil {
+			return errors.Wrap(err, "refreshSGIDs: failed to get local ENI metadata")
+		}
+
+		for _, eni := range allENIs {
+			eniID := eni.ENIID
+			log.Infof("GOT ENI %s", eniID)
+			// Also change the ENI's attribute so that the ENI will be deleted when the instance is deleted.
+			attributeInput := &ec2.ModifyNetworkInterfaceAttributeInput{
+				Groups:             sgIDsPtrs,
+				NetworkInterfaceId: aws.String(eniID),
+			}
+			start := time.Now()
+			_, err = cache.ec2SVC.ModifyNetworkInterfaceAttributeWithContext(context.Background(), attributeInput, userAgent)
+			awsAPILatency.WithLabelValues("ModifyNetworkInterfaceAttribute", fmt.Sprint(err != nil)).Observe(msSince(start))
+			if err != nil {
+				awsAPIErrInc("ModifyNetworkInterfaceAttribute", err)
+				return errors.Wrap(err, "refreshSGIDs: unable to update the ENI's SG")
+			}
+		}
+	}
```

Comment:
Please move this out of this method as this just to refresh/init security groups.

Reply:
@mogren - Discussed with Nithish. We wanted to know if this is can be done in the same thread or do you want to create a new thread something like "refreshsecuritygroup". This new thread would partially do the same operations as "refreshSGIDs" hence we wanted to check with you.